### PR TITLE
[runtime] Add an options API.

### PR DESCRIPTION
**Pull request by vargaz**
Add a general options API to the runtime, based on the flags API in Google V8:

```https://chromium.googlesource.com/v8/v8.git/+/refs/heads/master/src/flags/```

Supported features:
* Definition of runtime options in a declarative way.
* Options are mapped to C globals.
* BOOL/INT/STRING data types.
* Generic option parsing code.
* Generic usage code.
* Read-only flags for build-time optimization.

This is designed to replace the many option parsing functions in
the runtime, MONO_DEBUG, the many mono_set_... functions etc.

### CODE REVIEW

**Comment by vargaz** [comment]
Questions:
- maybe use the name 'option' instead of 'flag', flag implies a boolean.
- backward compatibility.

**Comment by lambdageek** [comment]
Nice.

Do you see this as a basis for new flags, or would we try to parse some of the existing ones using this mechanism?  (ie, do we need to add an arg to the macro to pick if it's a MONO_DEBUG flag or a GC flag or an AOT flag - and then have a bunch of additional parsing functions?)

I like the idea of using a regular prefix for them. I wonder if using some annoying prefix (`mono_FLAG_`) would encourage using getter functions instead of accessing the data directly.

**Comment by vargaz** [comment]
It can be used to parse some of the existing flags. I think its ok to access them directly, they should be read-only after startup so there is no need for synchronization etc., v8 does the same.

File: mono/utils/flags.c [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+			gboolean negate = FALSE;
+			if (!strcmp (flag->cmd_name, arg)) {
+			} else if (arg [0] == 'n' && arg [1] == 'o' && arg [2] == '-' &&
+					   !strcmp (flag->cmd_name, arg + 3)) {
+				negate = TRUE;
+			} else {
+				break;
+			}
+			if (flag->flag_type == MONO_FLAG_BOOL_READONLY) {
+				mono_error_set_error (error, 1, "Unable to set option '%s' as it's read only.\n", arg);
```
Comment (CoffeeFlux) [inline mono/utils/flags.c]:
```suggestion
				mono_error_set_error (error, 1, "Unable to set option '%s' as it's read-only.\n", arg);
```

File: mono/utils/flags.c [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+void
+mono_flags_print_usage (void)
+{
+#define DEFINE_FLAG_FULL(flag_type, ctype, c_name, cmd_name, def_value, comment) do { \
+		char *val = flag_value_to_str (flag_type, &mono_##c_name); \
+		g_printf ("  --%s (%s)\n\ttype: %s  default: %s\n", cmd_name, comment, flag_type_to_str (flag_type), val); \
+		g_free (val); \
+	} while (0);
+#include "flag-definitions.h"
```
Comment (CoffeeFlux) [inline mono/utils/flags.c]: I'm... not sure how I feel about this usage pattern, particularly in cases like this. I understand what's going on, but it took me a bit.

File: mono/utils/flags.c [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+static char *
+flag_value_to_str (MonoFlagType type, gconstpointer addr)
+{
+	switch (type) {
+	case MONO_FLAG_BOOL:
+	case MONO_FLAG_BOOL_READONLY:
+		return *(gboolean*)addr ? g_strdup ("true") : g_strdup ("false");
+	case MONO_FLAG_INT:
+		return g_strdup_printf ("%d", *(int*)addr);
```
Comment (CoffeeFlux) [inline mono/utils/flags.c]: Should probably move this function to goutput.c and make it behave with the stdout handler for the reasons explained above.

File: mono/utils/flags.c [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+	case MONO_FLAG_STRING:
+		return *(char**)addr ? g_strdup_printf ("%s", *(char**)addr) : g_strdup ("\"\"");
```
Comment (CoffeeFlux) [inline mono/utils/flags.c]: See above

File: mono/utils/flags.c [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+				g_hash_table_insert (flag_hash, (gpointer)flag_meta [i].cmd_name, &flag_meta [i]);
+			}
+		}
+
+		/* Compute flag name */
+		char *flagname = g_strdup (arg);
+		int len = strlen (arg);
+		int equals_sign_index = -1;
+		if (flagname [0] == 'n' && flagname [1] == 'o' && flagname [2] == '-') {
+			flagname += 3;
```
Comment (CoffeeFlux) [inline mono/utils/flags.c]: Update the length here as well? Also, what's even the point of this?

File: mono/utils/flags.c [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+		if (!flag) {
+			aindex ++;
+			continue;
+		}
+
+		switch (flag->flag_type) {
+		case MONO_FLAG_BOOL:
+		case MONO_FLAG_BOOL_READONLY: {
+			gboolean negate = FALSE;
+			if (!strcmp (flag->cmd_name, arg)) {
```
Comment (CoffeeFlux) [inline mono/utils/flags.c]: Store the length in the FlagData and check that first to avoid full string comparison?

File: mono/utils/flags.c [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+			}
+			*(gboolean*)flag->addr = negate ? FALSE : TRUE;
+			argv [aindex] = NULL;
+			break;
+		}
+		case MONO_FLAG_INT:
+		case MONO_FLAG_STRING: {
+			const char *value = NULL;
+
+			int flen = strlen (flag->cmd_name);
```
Comment (CoffeeFlux) [inline mono/utils/flags.c]: Same comment on length storage

File: mono/utils/flags.c [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+		if (flag_hash == NULL) {
+			/* Compute a hash to avoid n^2 behavior */
+			flag_hash = g_hash_table_new (g_str_hash, g_str_equal);
+			for (i = 0; i < G_N_ELEMENTS (flag_meta); ++i) {
+				g_hash_table_insert (flag_hash, (gpointer)flag_meta [i].cmd_name, &flag_meta [i]);
+			}
+		}
+
+		/* Compute flag name */
+		char *flagname = g_strdup (arg);
```
Comment (CoffeeFlux) [inline mono/utils/flags.c]: I don't think this is ever freed? Do you even need to duplicate here?

File: mono/utils/flags.c (hunk 5 above)
Comment (vargaz) [reply mono/utils/flags.c]: To enable a bool flag using --flag, and to disable it using --no-flag.

**Comment by CoffeeFlux** [comment]
> maybe use the name 'option' instead of 'flag', flag implies a boolean.
No strong opinion here.
> maybe add a prefix to the C variable names to make it clear what they are, like v8
uses FLAG_..., we could use mono_flag or mono_opt_.
That sounds good to me. Preferable to just cluttering up `mono_` directly.
> backward compatibility.
Can you be more specific?

File: mono/utils/flags.c (hunk 2 above)
Comment (lewurm) [reply mono/utils/flags.c]: We use that pattern quite a bit in the code base, examples:
* `#include "mono/cil/tables.def"`
* `#include "mini-ops.h"`
* `#include "mono/metadata/wrapper-types.h"`
* `#include "mintops.def"`
* and probably more.

Comment (vargaz) [reply mono/utils/flags.c]: Its kinda like how the HTML templating engines are used.

File: mono/utils/flags.c (hunk 5 above)
Comment (CoffeeFlux) [reply mono/utils/flags.c]: Oh, I see now.

File: mono/utils/flags.c (hunk 2 above)
Comment (CoffeeFlux) [reply mono/utils/flags.c]: Gotcha.

**Comment by vargaz** [comment]
As for the macros, the macro magic is restricted to flags.c mostly, the rest of the runtime only seems a bunch of C global variables.
As for backward compatibility, it would be nice to use this API for existing options, but those options are set in variety of ways, like the options in MonoDebugOptions can be set by calling mini_get_debug_options () and setting it through that.

File: mono/utils/flags.c [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+	MONO_FLAG_BOOL,
+	MONO_FLAG_BOOL_READONLY,
+	MONO_FLAG_INT,
+	MONO_FLAG_STRING
+} MonoFlagType;
+
+/* Define flags */
+#define DEFINE_FLAG_FULL(flag_type, ctype, c_name, cmd_name, def_value, comment) \
+	ctype mono_##c_name = def_value;
+#define DEFINE_FLAG_READONLY(flag_type, ctype, c_name, cmd_name, def_value, comment)
```
Comment (lewurm) [inline mono/utils/flags.c]: shouldn't it be this?
```suggestion
#define DEFINE_FLAG_READONLY(flag_type, ctype, c_name, cmd_name, def_value, comment) \
	const ctype mono_##c_name = def_value;
```

Comment (lewurm) [reply mono/utils/flags.c]: hmm, I just saw it's done in `flags.h`.  Not sure why those flags a defined again here.

Comment (vargaz) [reply mono/utils/flags.c]: No, its already defined in flags.h as a static const.

Comment (lewurm) [reply mono/utils/flags.c]: ah, it's due to `MONO_API_DATA`. nevermind 🙂

**Comment by lambdageek** [comment]
For the windows builds you need to update `msvc/libmonoutils-common.targets{,filters}`

**Comment by vargaz** [comment]
@monojenkins build failed

File: mono/utils/flag-definitions.h [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+ * The _READONLY variants generate C const variables so the compiler can optimize away their usage.
+ * Flag types:
+ * BOOL - gboolean
+ * INT - int
+ * STRING - (malloc-ed) char*
+ *
+ * Flags can be set on the command line using:
+ * --[no-]-flag (bool)
+ * --flag=value (int/string)
+ * --flag value (int/string)
```
Comment (jaykrell) [inline mono/utils/flag-definitions.h]: Please allow single leading dash, like autoconf?

**Comment by vargaz** [comment]
@monojenkins build Linux RISC-V

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@lambdageek what is the next step here? Is it still worth landing?

**Comment by lambdageek** [comment]
@vargaz I think we still want this in dotnet/runtime world, right? MONO_ENV_OPTIONS is still useful, and this is a way to uniformly define those.

@vargaz should probably remove the example flags before merging.  or disable them in a follow-up PR.

File: mono/utils/options.c [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+		g_assert_not_reached ();
+		return NULL;
+	}
+}
+
+void
+mono_options_print_usage (void)
+{
+#define DEFINE_OPTION_FULL(option_type, ctype, c_name, cmd_name, def_value, comment) do { \
+		char *val = option_value_to_str (option_type, &mono_##c_name); \
```
Comment (lambdageek) [inline mono/utils/options.c:81]:
```suggestion
		char *val = option_value_to_str (option_type, &mono_opt_##c_name); \
```

**Pull request merged**